drivers/gpio/gpio_psoc6.c

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright (c) 2018 Justin Watson
+ * Copyright (c) 2020 ATL Electronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT cypress_psoc6_gpio
+
+#include <errno.h>
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <soc.h>
+#include <drivers/gpio.h>
+
+#include "gpio_utils.h"
+#include "cy_gpio.h"
+#include "cy_sysint.h"
+
+#define LOG_LEVEL CONFIG_GPIO_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(gpio_psoc6);
+
+typedef void (*config_func_t)(const struct device *dev);
+
+struct gpio_psoc6_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	GPIO_PRT_Type *regs;
+	config_func_t config_func;
+};
+
+struct gpio_psoc6_runtime {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	sys_slist_t cb;
+};
+
+#define DEV_CFG(dev) \
+	((const struct gpio_psoc6_config * const)(dev)->config)
+#define DEV_DATA(dev) \
+	((struct gpio_psoc6_runtime * const)(dev)->data)
+
+static int gpio_psoc6_config(const struct device *dev,
+			     gpio_pin_t pin,
+			     gpio_flags_t flags)
+{
+	const struct gpio_psoc6_config * const cfg = DEV_CFG(dev);
+	GPIO_PRT_Type * const port = cfg->regs;
+	uint32_t drv_mode;
+	uint32_t pin_val;
+
+	if (flags & GPIO_OUTPUT) {
+		if (flags & GPIO_SINGLE_ENDED) {
+			drv_mode = (flags & GPIO_LINE_OPEN_DRAIN) ?
+				CY_GPIO_DM_OD_DRIVESLOW_IN_OFF :
+				CY_GPIO_DM_OD_DRIVESHIGH_IN_OFF;
+
+			pin_val = (flags & GPIO_LINE_OPEN_DRAIN) ? 1 : 0;
+		} else {
+			drv_mode = CY_GPIO_DM_STRONG_IN_OFF;
+
+			pin_val = (flags & GPIO_OUTPUT_INIT_HIGH) ? 1 : 0;
+		}
+	} else {
+		if ((flags & GPIO_PULL_UP) && (flags & GPIO_PULL_DOWN)) {
+			drv_mode = CY_GPIO_DM_PULLUP_DOWN_IN_OFF;
+		} else if (flags & GPIO_PULL_UP) {
+			drv_mode = CY_GPIO_DM_PULLUP_IN_OFF;
+		} else if (flags & GPIO_PULL_DOWN) {
+			drv_mode = CY_GPIO_DM_PULLDOWN_IN_OFF;
+		} else {
+			drv_mode = CY_GPIO_DM_ANALOG;
+		}
+
+		pin_val = (flags & GPIO_PULL_UP) ? 1 : 0;
+	}
+
+	if (flags & GPIO_INPUT) {
+		drv_mode |= CY_GPIO_DM_HIGHZ;
+	}
```

Comment:
This section is a bit confusing. I think it's better to remove it. In any case it will be optimized away by the compiler.

Reply:
I'm not sure if I can remove the section. Below is the **drv_mode** options: **0x01** is reserved
```
#define CY_GPIO_DM_ANALOG                      (0x00UL) /**< \brief Analog High-Z. Input buffer off */
#define CY_GPIO_DM_PULLUP_IN_OFF               (0x02UL) /**< \brief Resistive Pull-Up. Input buffer off */
#define CY_GPIO_DM_PULLDOWN_IN_OFF             (0x03UL) /**< \brief Resistive Pull-Down. Input buffer off */
#define CY_GPIO_DM_OD_DRIVESLOW_IN_OFF         (0x04UL) /**< \brief Open Drain, Drives Low. Input buffer off */
#define CY_GPIO_DM_OD_DRIVESHIGH_IN_OFF        (0x05UL) /**< \brief Open Drain, Drives High. Input buffer off */
#define CY_GPIO_DM_STRONG_IN_OFF               (0x06UL) /**< \brief Strong Drive. Input buffer off */
#define CY_GPIO_DM_PULLUP_DOWN_IN_OFF          (0x07UL) /**< \brief Resistive Pull-Up/Down. Input buffer off */

#define CY_GPIO_DM_HIGHZ                       (0x08UL) /**< \brief Digital High-Z. Input buffer on */
#define CY_GPIO_DM_PULLUP                      (0x0AUL) /**< \brief Resistive Pull-Up. Input buffer on */
#define CY_GPIO_DM_PULLDOWN                    (0x0BUL) /**< \brief Resistive Pull-Down. Input buffer on */
#define CY_GPIO_DM_OD_DRIVESLOW                (0x0CUL) /**< \brief Open Drain, Drives Low. Input buffer on */
#define CY_GPIO_DM_OD_DRIVESHIGH               (0x0DUL) /**< \brief Open Drain, Drives High. Input buffer on */
#define CY_GPIO_DM_STRONG                      (0x0EUL) /**< \brief Strong Drive. Input buffer on */
#define CY_GPIO_DM_PULLUP_DOWN                 (0x0FUL) /**< \brief Resistive Pull-Up/Down. Input buffer on */
```

I divided in two groups (without / with) input buffer. following above documentation.

Based on the valid configurations, when GPIO_INPUT is selected, I thought could be interesting combine any previous lines from **gpio_psoc6_config** method with **CY_GPIO_DM_HIGHZ**.

I hope was address your concern.

Reply:
Yes, thanks for the clarification. I've made my comment after reading the data sheet where HighZ drive mode is given value 0. I see now that this is not the case.

One more question. Considering the above defines shouldn't the code section in lines 67-73 responsible for configuring pull-up, pull-down behaviour be moved from the `else` clause to the `if (flags & GPIO_OUTPUT)` section? Seems like it's possible to configure pull-up, pull-down when output is enabled.

Reply:
> shouldn't the code section in lines 67-73 responsible for configuring pull-up, pull-down behaviour be moved from the `else` clause to the `if (flags & GPIO_OUTPUT)` section?

OK, I see. That wouldn't be appropriate since the output would not work in push-pull mode any more. The current implementation looks good. Thanks!